docs(web_socket_counter): create tutorial (#454)
VeryGoodOpenSource/dart_frog: docs/docs/tutorials/web_socket_counter.md

---
sidebar_position: 5
title: 🔌 WebSocket Counter
description: Build a real time "Counter" powered by WebSockets.
---

# WebSocket Counter 🔌

:::info
**Difficulty**: 🟠 Intermediate<br/>
**Length**: 30 minutes

> Before getting started, [read the prerequisites](/docs/overview#prerequisites) to make sure your development environment is ready.

Reviewer: ```suggestion
Before getting started, [read the Dart Frog prerequisites](/docs/overview#prerequisites) to make sure your development environment is ready.
```

:::

## Overview

> In this tutorial, we're going to build an app that exposes a single endpoint which handles WebSocket connections and maintains a real-time counter which can be incremented/decremented by connected clients.

Reviewer: ```suggestion
In this tutorial, we're going to build an app that exposes a single endpoint which handles WebSocket connections and maintains a real-time counter which can be incremented and decremented by connected clients.
```

> When we're done, we should be able to connect to the `/ws` endpoint and send/receive messages.

Reviewer: ```suggestion
When we're done, we should be able to connect to the `/ws` endpoint and send or receive messages.
```

```dart
import 'package:web_socket_channel/web_socket_channel.dart';

void main() async {
  final channel = WebSocketChannel.connect(Uri.parse('ws://localhost:8080/ws'));
  channel.stream.listen(print);

  channel.sink.add('__increment__');
  channel.sink.add('__decrement__');

  channel.sink.close();
}
```

> And we should see the following output:

Reviewer: ```suggestion
We should see the following output:
```

```
0 # initial
1 # increment
0 # decrement
```

## Creating a new app

> To create a new Dart Frog app, open your terminal, `cd` into the directory where you'd like to create the app, and run the following command:

Author: ```suggestion
To create a new Dart Frog app, open your terminal, change to the directory where you'd like to create the app, and run the following command:
```

```bash
dart_frog create web_socket_counter
```

You should see an output similar to:

```
✓ Creating web_socket_counter (0.1s)
✓ Installing dependencies (1.7s)

Created web_socket_counter at ./web_socket_counter.

> Get started by typing:

Reviewer: is this supposed to be markdown? I think we're missing some backticks

Author: No it's supposed to be the output from the terminal.

cd ./web_socket_counter
dart_frog dev
```

## Running the development server

> You should now have a directory called `web_socket_counter` -- `cd` into it:

Reviewer: ```suggestion
You should now have a directory called `web_socket_counter`. Let's `cd` into it:
```

> You should now have a directory called `web_socket_counter`. Let's `cd` into it:

Author: ```suggestion
You should now have a directory called `web_socket_counter`. Let's change directories into the newly created project:
```

cd web_socket_counter
```

Then, run the following command:

```bash
dart_frog dev
```

This will start the development server on port `8080`:

```
✓ Running on http://localhost:8080 (1.3s)
The Dart VM service is listening on http://127.0.0.1:8181/YKEF_nbwOpM=/
The Dart DevTools debugger and profiler is available at: http://127.0.0.1:8181/YKEF_nbwOpM=/devtools/#/?uri=ws%3A%2F%2F127.0.0.1%3A8181%2FYKEF_nbwOpM%3D%2Fws
[hotreload] Hot reload is enabled.
```

Make sure it's working by opening [http://localhost:8080](http://localhost:8080) in your browser or via `cURL`:

```bash
curl --request GET \
  --url http://localhost:8080
```

If everything succeeded, you should see `Welcome to Dart Frog!`.

## Creating the WebSocket Route

Now that we have a running application, let's start by creating a new `ws` route at `routes/ws.dart`:

```dart
import 'package:dart_frog/dart_frog.dart';

Response onRequest(RequestContext context) {
  return Response(body: 'You have requested /ws');
}
```

We can also delete the root endpoint at `routes/index.dart` since we won't be needing it for this example.

Save the changes and hot reload should kick in ⚡️

```
[hotreload] - Application reloaded.
```

> Now if we visit [http://localhost:8080/ws](http://localhost:8080/ws) in the browser or via cURL:

Reviewer: ```suggestion
Now if we visit [http://localhost:8080/ws](http://localhost:8080/ws) in the browser or via `cURL`:
```

  final channel = WebSocketChannel.connect(uri);

  // Listen to messages from the server.
  channel.stream.listen(print);

  // Send a message to the server.
  channel.sink.add('hello from the client');

  // Close the connection.
  channel.sink.close();
}
```

We're using [`package:web_socket_channel`](https://pub.dev/packages/web_socket_channel) to connect to our Dart Frog `/ws` endpoint. We can send messages to the server by calling `add` on the `WebSocketChannel` sink. We can listen to incoming messages by subscribing to the `WebSocketChannel` stream.

With the Dart Frog server still running, open a separate terminal, and run the example script:

```bash
dart example/main.dart
```

We should see the following output on the client:

```
hello from the server
```

On the server we should see the following output:

```
connected
hello from the client
disconnected
```

Awesome! We've configured a WebSocket handler and established a connection to our server 🎉

## Managing the Counter State

Now that we've configured the WebSocket handler, we're going to shift gears and work on creating a component that will manage the state of the counter.

> In this example, we're going to use a cubit from the [Bloc Library](https://bloclibrary.dev) to manage the state of our counter because it provides a reactive API which allows us to stream state changes and query the current state at any given point in time. We're going to use [package:broadcast_bloc](https://pub.dev/packages/broadcast_bloc) which allows blocs or cubits to broadcast their state changes to any subscribed stream channels -- this will come in handy later on.

Reviewer: ```suggestion
In this example, we're going to use a cubit from the [Bloc Library](https://bloclibrary.dev) to manage the state of our counter because it provides a reactive API which allows us to stream state changes and query the current state at any given point in time. We're going to use [package:broadcast_bloc](https://pub.dev/packages/broadcast_bloc) which allows blocs or cubits to broadcast their state changes to any subscribed stream channels — this will come in handy later on.
```

```
dart pub add broadcast_bloc
```

Then, create a cubit in `lib/counter/cubit/counter_cubit.dart`.

```dart
import 'package:broadcast_bloc/broadcast_bloc.dart';

class CounterCubit extends BroadcastCubit<int> {
  // Create an instance with an initial state of 0.
  CounterCubit() : super(0);

  // Increment the current state.
  void increment() => emit(state + 1);

  // Decrement the current state.
  void decrement() => emit(state - 1);
}
```

In order to access the cubit from our route handler, we'll create a `provider` in `lib/counter/middleware/counter_provider.dart`.

```dart
import 'package:dart_frog/dart_frog.dart';
import 'package:web_socket_counter/counter/counter.dart';

final _counter = CounterCubit();

// Provide the counter instance via `RequestContext`.
final counterProvider = provider<CounterCubit>((_) => _counter);
```

:::info
For more information, refer to the [dependency injection documentation](/docs/basics/dependency-injection).
:::

Let's also create a barrel file which exports all `counter` components in `lib/counter/counter.dart`:

```dart
export 'cubit/counter_cubit.dart';
export 'middleware/counter_provider.dart';
```

## Providing the Counter

> We need to use the `counterProvider` in order to have access to it in sub-routes. Create a global piece of middleware (`routes/_middleware.dart`):

Author: ```suggestion
We need to use the `counterProvider` in order to have access to it in nested. Create a global piece of middleware (`routes/_middleware.dart`):
```

> We need to use the `counterProvider` in order to have access to it in sub-routes. Create a global piece of middleware (`routes/_middleware.dart`):

Reviewer: ```suggestion
We need to use the `counterProvider` in order to have access to it in nested routes. Create a global piece of middleware (`routes/_middleware.dart`):
```

      );
    },
  );

  return handler(context);
}
```

First, we subscribe the newly connected client to the `CounterCubit` in order to receive updates whenever the cubit state changes.

Next, we send the current count to the new client via `cubit.state`.

When the client sends a new message, we invoke the `increment`/`decrement` method on the cubit based on the message.

Finally, we unsubscribe the channel when the client disconnects.

:::info
The `subscribe` and `unsubscribe` APIs are exposed by the `BroadcastCubit` super class from `package:broadcast_bloc`.
:::

Be sure to save all the changes and hot reload should kick in ⚡️

```
[hotreload] - Application reloaded.
```

Now we can update our example script in `example/main.dart`:

```dart
import 'package:web_socket_channel/web_socket_channel.dart';

void main() async {
  final channel = WebSocketChannel.connect(Uri.parse('ws://localhost:8080/ws'));
  channel.stream.listen(print);

  channel.sink.add('__increment__');
  channel.sink.add('__decrement__');

  channel.sink.close();
}
```

Finally, let's run the script:

```
dart example/main.dart
```

We should see the following output:

```
0
1
0
```

:::note

> If you restart the server, the count will always be reset to 0 because it is only maintained in memory.

Reviewer: ```suggestion
If you restart the server, the count will always be reset to 0 because it is only maintained in memory.
```
is this because we're running the example locally?

Author: No it's just because the count isn't being persisted anywhere for simplicity.

}
```

We're using [`package:web_socket_channel`](https://pub.dev/packages/web_socket_channel) to connect to our Dart Frog `/ws` endpoint. We can send messages to the server by calling `add` on the `WebSocketChannel` sink. We can listen to incoming messages by subscribing to the `WebSocketChannel` stream.

> With the Dart Frog server still running, open a separate terminal and run the example script:

Reviewer: ```suggestion
With the Dart Frog server still running, open a separate terminal, and run the example script:
```

Now that we've configured the WebSocket handler, we're going to shift gears and work on creating a component that will manage the state of the counter.

> In this example, we're going to use a cubit from the [Bloc Library](https://bloclibrary.dev) to manage the state of our counter because it provides a reactive API which allows us to stream state changes and query the current state at any given point in time. We're going to use [package:broadcast_bloc](https://pub.dev/packages/broadcast_bloc) which allows blocs/cubits to broadcast their state changes to any subscribed stream channels -- this will come in handy later on.

Reviewer: ```suggestion
In this example, we're going to use a cubit from the [Bloc Library](https://bloclibrary.dev) to manage the state of our counter because it provides a reactive API which allows us to stream state changes and query the current state at any given point in time. We're going to use [package:broadcast_bloc](https://pub.dev/packages/broadcast_bloc) which allows blocs or cubits to broadcast their state changes to any subscribed stream channels -- this will come in handy later on.
```

Reviewer: Should we add a :::note about broadcast_bloc and what it is?

> 

Reviewer: nested routes instead of sub-routes plz
